### DISCUSSION TITLE
[cloud_firestore] Make Futures work in offline-mode

### DISCUSSION THREAD

**Discussion by oodavid** [firebase/flutterfire#7002] [Feature request]
The docs and examples recommend making use of `Futures`, but they remain uncompleted when offline. This will be unexpected for most users.

Offline behaviour is described as using a local database, with server sync happening in the background:

> Firestore provides out of the box support for offline capabilities. When reading and writing data, Firestore uses a local database which automatically synchronizes with the server. Cloud Firestore functionality continues when users are offline, and automatically handles data migration when they regain connectivity.
>
> Source: https://firebase.flutter.dev/docs/firestore/usage#access-data-offline

I know this is described as [intended behaviour](https://github.com/FirebaseExtended/flutterfire/issues/3842#issuecomment-708399905), however it makes using the Futures unusable for apps that use offline behaviour with `write` operations.

## Proposal

Allow us to decide how Futures are completed. On a global, or per-query basis. The `Source` enum springs to mind:

```dart
// Default behaviour, the future completes after being written to the server
docRef.set(
  data,
  SetOptions(completionFrom: Source.server),
); 

// The future completes after being written to the cache
docRef.set(
  data,
  SetOptions(completionFrom: Source.cache),
);

// The future completes after being written to the server AND cache
docRef.set(
  data,
  SetOptions(completionFrom: Source.serverAndCache),
);
```

That way we can cater for existing behaviour with defaults, and also improve the offline functionality.

## Video

I originally thought this was a bug. It's since change to a feature request. Here's my original:

https://www.loom.com/share/30f4f0fa6d184ebe8df1751c526b2f4e

**Comment by pbouttier** [comment]
Hello @oodavid ,

I've encountered the same problem recently, and I haven't found a solution, either by searching here or on the android sdk directly. 
Have you found a solution or, failing that, a good “safe” implementation?

Thanks
